zh-cn: init translation of Svelte stores (#17974)
mdn/translated-content: files/zh-cn/learn/tools_and_testing/client-side_javascript_frameworks/svelte_stores/index.md

   更新 `stores.js` 如下所示：

   ```js
   import { writable } from "svelte/store";
   import { localStore } from "./localStore.js";

   export const alert = writable("Welcome to the to-do list app!");

   const initialTodos = [
     { id: 1, name: "Visit MDN web docs", completed: true },
     { id: 2, name: "Complete the Svelte Tutorial", completed: false },
   ];

   export const todos = localStore("mdn-svelte-todo", initialTodos);
   ```

   通过使用 `localStore('mdn-svelte-todo', initialTodos)`，我们配置 store 将数据保存在 Web 存储中的键 `mdn-svelte-todo` 下。我们还设置了一些待办事项作为初始值。

4. 现在让我们删除 `App.svelte` 中的硬编码待办事项。更新它的内容如下所示。我们基本上只是删除了 `$todos` 数组和 `console.log()` 语句：

   ```svelte
   <script>
     import Todos from './components/Todos.svelte'
     import Alert from './components/Alert.svelte'

     import { todos } from './stores.js'
   </script>

   <Alert />
   <Todos bind:todos={$todos} />
   ```

   > **备注：** 这是我们为了使用自定义存储所必须进行的唯一更改。`App.svelte` 完全不需要知道我们使用哪种类型的 store。

5. 现在再次运行你的应用程序。创建一些待办事项，然后关闭浏览器。你甚至可以停止 Svelte 服务器并重新启动它。当你重新访问该 URL 时，你的待办事项仍将存在。
6. 你还可以在开发者工具的控制台中进行检查。在 Web 控制台中，输入命令 `localStorage.getItem('mdn-svelte-todo')`。对应用程序进行一些更改，比如按下“全部取消”按钮，然后再次检查 Web 存储的内容。你将会看到类似以下的内容：

>    ![带有 Web 控制台视图的待办事项应用程序，显示当在应用程序中更改待办事项时，相应的条目也在 Web 存储中更改](03-persisting-todos-to-local-storage.png)
> Svelte store 提供了一种非常简单、轻量但功能强大的方式，以响应式的方式处理来自全局数据存储的复杂应用程序状态。由于 Svelte 编译我们的代码，它可以提供 [`$store` 自动订阅语法](https://svelte.dev/docs/svelte-components#script-4-prefix-stores-with-$-to-access-their-values)，使我们可以通过与本地变量相同的方式处理存储。由于 store 具有很简单的 API，因此创建我们自己的自定义 store 来抽象 store 本身的内部工作方式非常简单。

Author: ```suggestion
   ![带有 Web 控制台视图的待办事项应用程序，显示当在应用程序中更改待办事项时，相应的条目也在 Web 存储中更改](03-persisting-todos-to-local-storage.png)

Svelte store 提供了一种非常简单、轻量但功能强大的方式，以响应式的方式处理来自全局数据存储的复杂应用程序状态。由于 Svelte 编译我们的代码，它可以提供 [`$store` 自动订阅语法](https://svelte.dev/docs/svelte-components#script-4-prefix-stores-with-$-to-access-their-values)，使我们可以通过与本地变量相同的方式处理存储。由于 store 具有很简单的 API，因此创建我们自己的自定义 store 来抽象 store 本身的内部工作方式非常简单。
```

---
title: 使用 Svelte store
slug: Learn/Tools_and_testing/Client-side_JavaScript_frameworks/Svelte_stores

> ---

Reviewer: 建议参考 https://github.com/mdn/translated-content/blob/main/docs/zh-cn/translation-guide.md#%E5%85%83%E6%95%B0%E6%8D%AE 添加 l10n.sourceCommit

> slug: Learn/Tools_and_testing/Client-side_JavaScript_frameworks/Svelte_stores
> ---

Author: ```suggestion
slug: Learn/Tools_and_testing/Client-side_JavaScript_frameworks/Svelte_stores
l10n:
  sourceCommit: e9be22eaa7416206e3f263a058f0d509a7f81f88
---
```

{{LearnSidebar}}
{{PreviousMenuNext("Learn/Tools_and_testing/Client-side_JavaScript_frameworks/Svelte_reactivity_lifecycle_accessibility","Learn/Tools_and_testing/Client-side_JavaScript_frameworks/Svelte_TypeScript", "Learn/Tools_and_testing/Client-side_JavaScript_frameworks")}}

在上一篇文章中，我们完成了应用程序的开发，将其组织为组件，并讨论了一些处理响应式、处理 DOM 节点以及暴露组件功能的高级技术。在本文中，我们将展示另一种在 Svelte 中处理状态管理的方式：[store](https://learn.svelte.dev/tutorial/writable-stores)。store 是全局数据存储库，用于保存值。组件可以订阅 store 并在其值发生变化时接收通知。

<table>
  <tbody>
    <tr>
      <th scope="row">前提：</th>
      <td>
        <p>
          至少，建议你熟悉核心的
          <a href="/zh-CN/docs/Learn/HTML">HTML</a>、
          <a href="/zh-CN/docs/Learn/CSS">CSS</a> 和
          <a href="/zh-CN/docs/Learn/JavaScript">JavaScript</a>

>           语言，并且了解<a
>             href="/zh-CN/docs/Learn/Tools_and_testing/Understanding_client-side_tools/Command_line"
>             >终端/命令行</a
>           >的使用方法。

Reviewer: ```suggestion
          语言，并且了解
          <a href="/zh-CN/docs/Learn/Tools_and_testing/Understanding_client-side_tools/Command_line">终端/命令行</a>
          的使用方法。
```

Reviewer 2: 他的布局是正确的（中文之间要避免引入换行带来的空格）

Reviewer: got it!

>           至少，建议你熟悉核心的
>           <a href="/zh-CN/docs/Learn/HTML">HTML</a>、
>           <a href="/zh-CN/docs/Learn/CSS">CSS</a> 和
>           <a href="/zh-CN/docs/Learn/JavaScript">JavaScript</a>
>           语言，并且了解<a
>             href="/zh-CN/docs/Learn/Tools_and_testing/Understanding_client-side_tools/Command_line"
>             >终端/命令行</a
>           >的使用方法。

Author: ```suggestion
          至少，建议你熟悉核心的
          <a href="/zh-CN/docs/Learn/HTML">HTML</a>、<a
               href="/zh-CN/docs/Learn/CSS">CSS</a> 和
          <a href="/zh-CN/docs/Learn/JavaScript">JavaScript</a>
          语言，并且了解<a
            href="/zh-CN/docs/Learn/Tools_and_testing/Understanding_client-side_tools/Command_line"
            >终端/命令行</a
          >的使用方法。
```

        </p>
        <p>
          你需要安装了 node 和 npm 的终端，以编译和构建你的应用程序。
        </p>
      </td>
    </tr>
    <tr>
      <th scope="row">目标：</th>
      <td>学习如何使用 Svelte store</td>
    </tr>
  </tbody>
</table>

我们将使用 store，创建名为 `Alert` 的组件，用于在屏幕上显示通知。这个组件可以从任何组件中接收消息。在这种情况下，`Alert` 组件是独立于其他组件的——它不是任何其他组件的父组件或子组件，因此消息不符合组件层次结构。

我们还将看到如何开发自己的自定义 store，将待办事项信息持久化到 [Web 存储](/zh-CN/docs/Web/API/Web_Storage_API)中，使我们的待办事项在页面重新加载后不变。

## 与我们一起编写代码

### Git

使用以下命令克隆 GitHub 仓库（如果尚未完成）：

```bash
git clone https://github.com/opensas/mdn-svelte-tutorial.git
```

然后进入当前应用程序状态的目录：

```bash
cd mdn-svelte-tutorial/06-stores
```

或者直接下载文件夹的内容：

```bash
npx degit opensas/mdn-svelte-tutorial/06-stores
```

记得运行 `npm install && npm run dev` 以开发模式启动应用程序。

### REPL

要使用 REPL 和我们一起编写程序代码，点击如下链接：

<https://svelte.dev/repl/d1fa84a5a4494366b179c87395940039?version=3.23.2>

## 处理应用程序状态

> 我们已经学习如何使用属性、双向数据绑定和事件使组件之间进行通信。在所有这些情况下，我们处理的都是父子组件之间的通信。

Reviewer: ```suggestion
我们已经学习如何使用属性、双向数据绑定和事件在组件之间进行通信。在所有这些情况下，我们处理的都是父子组件之间的通信。
```

但并不是所有的应用程序状态都应该放在应用程序的组件层次结构中。例如，关于已登录用户的信息，或者是否选择了暗黑主题。

有时，你的应用程序状态需要被多个没有层次关系的组件访问，或者被常规的 JavaScript 模块访问。

此外，当你的应用程序变得复杂，组件层次结构变得复杂时，组件之间传递数据可能变得非常困难。在这种情况下，转向全局数据存储可能是一个不错的选择。如果你已经使用过 [Redux](https://redux.js.org/) 或 [Vuex](https://vuex.vuejs.org/)，那么你应该很熟悉这种存储方式。Svelte store 提供了类似的状态管理特性。

> store 是带有 `subscribe()` 方法的对象，允许感兴趣的方在它的值变化时得到通知，并且还可以有一个可选的 `set()` 方法，用于设置 store 的新值。这套最小化的 API 被称为 [store contract](https://svelte.dev/docs/svelte-components#script-4-prefix-stores-with-$-to-access-their-values-store-contract)。

Reviewer: ```suggestion
store 是带有 `subscribe()` 方法的对象，允许感兴趣的一方在它的值变化时得到通知，并且还可以有一个可选的 `set()` 方法，用于设置 store 的新值。这套最小化的 API 被称为 [store contract](https://svelte.dev/docs/svelte-components#script-4-prefix-stores-with-$-to-access-their-values-store-contract)。
```

> store 是带有 `subscribe()` 方法的对象，允许感兴趣的一方在它的值变化时得到通知，并且还可以有一个可选的 `set()` 方法，用于设置 store 的新值。这套最小化的 API 被称为 [store contract](https://svelte.dev/docs/svelte-components#script-4-prefix-stores-with-$-to-access-their-values-store-contract)。

Author: ```suggestion
store 是带有 `subscribe()` 方法的对象，允许感兴趣的一方在它的值变化时得到通知，并且还可以有一个可选的 `set()` 方法，用于设置 store 的新值。这套最小化的 API 被称为 [store 合约](https://svelte.dev/docs/svelte-components#script-4-prefix-stores-with-$-to-access-their-values-store-contract)。
```

svelte 在 `svelte/store` 模块中提供了创建[可读](https://svelte.dev/docs/svelte-store#readable)、[可写](https://svelte.dev/docs/svelte-store#writable)和[派生](https://svelte.dev/docs/svelte-store#derived) store 的函数。

> svelte 还提供了一种非常直观的将 store 集成到其响应式系统中的方式：[响应式 `$store` 语法](https://svelte.dev/docs/svelte-components#script-4-prefix-stores-with-$-to-access-their-values)。如果你按照 store contract 创建自己的 store，你将获得这种响应式的语法糖而无需多余代码。

Author: ```suggestion
svelte 还提供了一种非常直观的将 store 集成到其响应式系统中的方式：[响应式 `$store` 语法](https://svelte.dev/docs/svelte-components#script-4-prefix-stores-with-$-to-access-their-values)。如果你按照 store  合约创建自己的 store，你将获得这种响应式的语法糖而无需多余代码。
```

> svelte 还提供了一种非常直观的将 store 集成到其响应式系统中的方式：[响应式 `$store` 语法](https://svelte.dev/docs/svelte-components#script-4-prefix-stores-with-$-to-access-their-values)。如果你按照 store contract 创建自己的 store，你将获得这种响应式的语法糖而无需多余代码。

Author: ```suggestion
svelte 还提供了一种非常直观的将 store 集成到其响应式系统中的方式：[响应式 `$store` 语法](https://svelte.dev/docs/svelte-components#script-4-prefix-stores-with-$-to-access-their-values)。如果你按照 store 合约创建自己的 store，你将获得这种响应式的语法糖而无需多余代码。
```

## 创建 Alert 组件

为了展示如何使用 store，我们将创建 `Alert` 组件。这种小部件也被称为弹出通知、toast 或者提示气泡。

我们的 `Alert` 组件将由 `App` 组件显示，但任何组件都可以向其发送通知。每当有通知到达时，`Alert` 组件将负责在屏幕上显示它。

### 创建 store

让我们从创建一个可写 store 开始。任何组件都可以向这个 store 写入数据，`Alert` 组件将订阅它，并在 store 修改时显示消息。

1. 在你的 `src` 目录下创建一个名为 `stores.js` 的新文件。
2. 向新文件填入以下内容：

   ```js
   import { writable } from "svelte/store";

   export const alert = writable("Welcome to the to-do list app!");
   ```

> **备注：** store 可以在 Svelte 组件之外定义和使用，所以你可以按照自己的方式组织它们。

在上面的代码中，我们从 `svelte/store` 导入 `writable()` 函数，并使用它创建了名为 `alert` 的新 store，初始值为“Welcome to the to-do list app!”（欢迎来到待办事项列表应用程序！）。然后我们通过 `export` 导出该 store。

> 

Reviewer: 建议直接翻译代码的 “Welcome to the to-do list app!” 为中文

下同

Reviewer 2: 因为引用的外部资源未被本地化，所以对于这些关联外部示例代码的内容，我们暂时不做本地化

Reviewer: soga

     <p>{ $alert }</p>
   </div>
   {/if}
   ```

这里我们首先创建了默认值为 3000（毫秒）的属性 `ms`。然后，我们创建了 `onMessageChange()` 函数，它负责控制 Alert 是否可见。通过 `$: onMessageChange($alert, ms)`，我们告诉 Svelte 在 `$alert` store 或 `ms` 属性发生变化时运行此函数。

每当 `$alert` store 发生变化时，我们将清除任何等待的超时。如果 `$alert` 为空，我们将 `visible` 设置为 `false`，之后 `Alert` 会被从 DOM 中删除。如果不为空，我们将 `visible` 设置为 `true`，并使用 `setTimeout()` 函数以在 `ms` 毫秒后清除提示。

最后，使用 `onDestroy()` 生命周期函数，确保调用 `clearTimeout()` 函数。

我们还在提示段落上方添加了 SVG 图标，使其看起来更漂亮。再次尝试一下，你应该会看到这些更改。

## 使我们的 Alert 组件具有无障碍性

我们的 `Alert` 组件工作得很好，但对辅助技术来说并不友好。问题在于页面上动态添加和删除的元素。虽然对于能看到页面的用户来说很明显，但对于使用辅助技术（如屏幕阅读器）的用户来说可能不太明显。为了处理这些情况，我们可以利用 [ARIA 实时区域](/zh-CN/docs/Web/Accessibility/ARIA/ARIA_Live_Regions)，它提供了以编程方式公开动态内容更改的方法，以便辅助技术可以检测并宣布这些更改。

> 我们可以使用 `aria-live` 属性和礼貌设置来声明包含动态内容的区域，以便辅助技术可以宣布它们。礼貌设置用于设置屏幕阅读器处理该区域更新的优先级。可能的设置有 `off`、`polite` 或 `assertive`。

Author: > “politeness setting”的翻译有出处吗

No, there is no source for the translation of it

> 我们可以使用 `aria-live` 属性和礼貌设置来声明包含动态内容的区域，以便辅助技术可以宣布它们。礼貌设置用于设置屏幕阅读器处理该区域更新的优先级。可能的设置有 `off`、`polite` 或 `assertive`。

Reviewer 2: 我倾向于使用“礼让设置”这一翻译，根据 [`aria-live` 文档](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions#live_regions)，该设置用于配置优先级，与 [Google 设置广告载入优先级的文档](https://support.google.com/richmedia/answer/2672514?hl=zh-Hans)类似。

/cc @jasonren0403

对于常见情况，你还可以使用几个预定义的专门的 `role` 值，如 `log`、`status` 和 `alert`。

在我们的情况下，只需将 `role="alert"` 添加到 `<div>` 容器中，就可以解决问题，如下所示：

```svelte
<div role="alert" on:click={() => visible = false}>
```

通常情况下，使用屏幕阅读器测试应用程序是一个好主意，不仅可以发现可访问性问题，还可以熟悉视觉障碍人士如何使用 Web。你有几个选项，例如 Windows 上的 [NVDA](https://www.nvaccess.org/)、Chrome 上的 [ChromeVox](https://support.google.com/chromebook/answer/7031755)、Linux 上的 [Orca](https://wiki.gnome.org/Projects/Orca)、macOS 和 iOS 上的 [VoiceOver](https://www.apple.com/accessibility/vision/)，以及其他选项。

要了解有关检测和修复无障碍问题的更多信息，请查看我们的[解决常见的无障碍问题](/zh-CN/docs/Learn/Tools_and_testing/Cross_browser_testing/Accessibility)文章。

> ## 使用存储合约持久化我们的待办事项

Author: ```suggestion
## 使用 store 合约持久化我们的待办事项
```

> ## 使用存储合约持久化我们的待办事项

Reviewer 2: 上文未翻译“store contract”，请修正上文的翻译（可以在第一处出现的地方添加英文注释）

3. 返回到你的应用程序，尝试添加一些待办事项，然后转到开发者工具的 Web 控制台。你会看到，我们对待办事项进行的每个修改都会通过 `bind` 指令反映在 `App.svelte` 中定义的 `todos` 数组中。

现在我们需要找到一种方法来持久化这些待办事项。我们可以在 `App.svelte` 组件中实现一些代码，将待办事项读取并保存到 [Web 存储](/zh-CN/docs/Web/API/Web_Storage_API)或 Web 服务中。但是，如果我们能开发出通用的存储机制来保存其内容，岂不是更好？这样我们就可以像使用其他存储一样使用它，并将持久化机制进行抽象。我们可以创建一个将内容同步到 Web 存储的 store，然后再开发另一个与 Web 服务同步的 store。在它们之间进行切换将变得非常简单，而且不需要修改 `App.svelte` 的任何代码。

### 保存待办事项

让我们首先使用常规的可写 store 来保存我们的待办事项。

1. 打开 `stores.js` 文件，在现有存储下面添加以下存储：

   ```js
   export const todos = writable([]);
   ```

2. 这很简单。现在我们需要导入该 store 并在 `App.svelte` 中使用它。只需记住，现在要访问待办事项，我们需要使用 `$todos` 以及响应式的 `$store` 语法。

   更新你的 `App.svelte` 文件如下：

   ```svelte
   <script>
     import Todos from "./components/Todos.svelte";
     import Alert from "./components/Alert.svelte";

     import { todos } from "./stores.js";

     $todos = [
       { id: 1, name: "Create a Svelte starter app", completed: true },
       { id: 2, name: "Create your first component", completed: true },
       { id: 3, name: "Complete the rest of the tutorial", completed: false }
     ];
   </script>

   <Alert />
   <Todos bind:todos={$todos} />
   ```

3. 尝试一下，一切应该正常工作。接下来，我们将看到如何定义自己的自定义存储。

> ### 如何实现存储合约：理论

Author: ```suggestion
### 如何实现 store 合约：理论
```

> 你可以通过实现存储合约来创建自己的 store，而无需依赖于 `svelte/store`。它必须按照以下方式工作：

Author: ```suggestion
你可以通过实现 store 合约来创建自己的 store，而无需依赖于 `svelte/store`。它必须按照以下方式工作：
```

1. store 必须包含 `subscribe()` 方法，该方法必须接受订阅函数作为参数。每当 store 的值发生更改时，必须调用所有订阅函数。
2. `subscribe()` 方法必须返回 `unsubscribe()` 函数，当调用 `unsubscribe()` 函数时，必须停止订阅。
3. store 可以选择包含 `set()` 方法，该方法必须接受 store 的新值作为参数，并同步调用所有订阅函数。具有 `set()` 方法的 store 称为可写 store。

首先，让我们向 `App.svelte` 组件添加以下 `console.log()` 语句，以查看 `todos` store 及其内容的工作情况。将这些行添加到 `todos` 数组的下方：

```js
console.log("todos store - todos:", todos);
console.log("todos store content - $todos:", $todos);
```

现在运行应用程序，你将在 Web 控制台中看到类似于以下内容：

![在 Web 控制台中显示 todos store 的函数和内容](02-svelte-store-in-action.png)

如你所见，我们的 store 只是一个包含 `subscribe()`、`set()` 和 `update()` 方法的对象，而 `$todos` 则是我们的待办事项数组。

> 以下是一个基本的可工作的 store，从头开始实现：

Reviewer 2: ```suggestion
以下是一个（从头开始实现的）基本的可工作的 store：
```

```js
export const writable = (initial_value = 0) => {
  let value = initial_value; // store 的内容

>   let subs = []; // 订阅者的处理函数

Reviewer 2: ```suggestion
  let subs = []; // 订阅者的处理器
```

  const subscribe = (handler) => {

>     subs = [...subs, handler]; // 将处理函数添加到订阅者数组中

Reviewer 2: ```suggestion
    subs = [...subs, handler]; // 将处理器添加到订阅者数组中
```

    handler(value); // 用当前值调用处理函数
    return () => (subs = subs.filter((sub) => sub !== handler)); // 返回取消订阅的函数
  };

  const set = (new_value) => {
    if (value === new_value) return; // 相同的值，退出
    value = new_value; // 更新值
    subs.forEach((sub) => sub(value)); // 更新订阅者
  };

  const update = (update_fn) => set(update_fn(value)); // 更新函数

  return { subscribe, set, update }; // 存储器的接口
};
```

> 在这里，我们声明了 `subs`，它是订阅者数组。在 `subscribe()` 方法中，我们将处理函数添加到 `subs` 数组中，并返回一个函数，当执行该返回的函数时，它将从数组中删除该处理函数。

Reviewer 2: ```suggestion
在这里，我们声明了 `subs`，它是订阅者数组。在 `subscribe()` 方法中，我们将处理器添加到 `subs` 数组中，并返回一个在执行时会从数组中删除该处理处理器的函数。
```

> 当我们调用 `set()` 时，我们更新 store 的值并调用每个处理函数，在这里将新值作为参数传递。

Reviewer 2: ```suggestion
当我们调用 `set()` 时，我们更新 store 的值并将新值作为参数来调用每个处理器。
```

通常情况下，你不需要从头开始实现 store；相反，你可以使用可写 store 来创建具有特定逻辑的[自定义 store](https://learn.svelte.dev/tutorial/custom-stores)。在以下示例中，我们创建了计数器 store，它只允许我们将计数器加一或重置其值：

```js
import { writable } from "svelte/store";
function myStore() {
  const { subscribe, set, update } = writable(0);

  return {
    subscribe,
    addOne: () => update((n) => n + 1),
    reset: () => set(0),
  };
}
```

> 如果我们的待办事项列表应用程序变得太复杂，我们可以让待办事项 store 处理所有状态修改。我们可以将所有修改 `todo` 数组的方法（如 `addTodo()`、`removeTodo()` 等）从我们的 `Todos` 组件移动到 store 中。如果有一个集中处理所有状态修改的地方，组件只需调用这些方法来修改应用程序的状态，并以响应式的方式显示 store 暴露的信息。有一个统一的处理状态修改的地方可以更容易地理解状态流动并发现问题。

Reviewer 2: ```suggestion
如果我们的待办事项列表应用程序变得太复杂，我们可以让待办事项 store 处理所有状态修改。我们可以将所有修改 `todo` 数组的方法（如 `addTodo()`、`removeTodo()` 等）从我们的 `Todos` 组件移动到 store 中。如果有一个集中处理所有状态修改的地方，组件只需调用这些方法来修改应用程序的状态，并以响应式的方式显示 store 暴露的信息。有一个统一的处理状态修改的地方可以更容易地推理状态流并发现问题。
```

Svelte 不会强制你以特定的方式组织状态管理；它只提供了工具，你可以自由选择如何处理它。

### 实现自定义的待办事项 store

我们的待办事项列表应用程序并不特别复杂，因此我们不会将所有的修改方法移到中心位置。我们将保持它们原样，集中精力将待办事项持久化。

> > **备注：** 如果你正在 Svelte REPL 上遵循本指南进行操作，你将无法完成此步骤。出于安全原因，Svelte REPL 在沙箱环境中工作，不能让你访问 Web 存储，你将收到“The operation is insecure”错误。要遵循本部分，你需要克隆存储库并转到 `mdn-svelte-tutorial/06-stores` 文件夹，或者你可以使用 `npx degit opensas/mdn-svelte-tutorial/06-stores` 直接下载该文件夹的内容。

Reviewer 2: ```suggestion
> **备注：** 如果你正在 Svelte REPL 上遵循本指南进行操作，你将无法完成此步骤。出于安全原因，Svelte REPL 在沙箱环境中工作，不允许你访问 Web 存储，你将收到“The operation is insecure”错误。要跟着完成本部分，你需要克隆存储库并转到 `mdn-svelte-tutorial/06-stores` 文件夹，或者你可以使用 `npx degit opensas/mdn-svelte-tutorial/06-stores` 直接下载该文件夹的内容。
```

> 要实现将其内容保存到 Web 存储的自定义 store，我们需要做以下内容的可写 store：

Reviewer 2: ```suggestion
要实现将其内容保存到 Web 存储的自定义 store，我们需要一个执行以下操作的可写 store：
```

   - 我们使用 [`localStorage.getItem(key)`](/zh-CN/docs/Web/API/Storage/getItem) 和 [`localStorage.setItem(key, value)`](/zh-CN/docs/Web/API/Storage/setItem) 方法来读取和写入 Web 存储中的信息，使用 [`toString()`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Object/toString) 和 `toObj()`（使用 [`JSON.parse()`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse)）辅助函数来转换值。
   - 接下来，我们将从 Web 存储中接收到的字符串内容转换为对象，并将该对象保存在我们的 store 中。
   - 最后，每当我们更新 store 的内容时，我们也会更新 Web 存储，Web 存储的新值为 store 的内容转换成的字符串。

   请注意，我们只需要重新定义 `set()` 方法，添加将值保存到 Web 存储的操作。其余的代码主要是初始化和转换相关的工作。

3. 现在，我们将在 `stores.js` 中使用本地存储来创建我们的本地持久化的待办事项 store。

   更新 `stores.js` 如下所示：

   ```js
   import { writable } from "svelte/store";
   import { localStore } from "./localStore.js";

   export const alert = writable("Welcome to the to-do list app!");

   const initialTodos = [
     { id: 1, name: "Visit MDN web docs", completed: true },
     { id: 2, name: "Complete the Svelte Tutorial", completed: false },
   ];

   export const todos = localStore("mdn-svelte-todo", initialTodos);
   ```

   通过使用 `localStore('mdn-svelte-todo', initialTodos)`，我们配置 store 将数据保存在 Web 存储中的键 `mdn-svelte-todo` 下。我们还设置了一些待办事项作为初始值。

> 4. 现在让我们删除 `App.svelte` 中的硬编码待办事项。更新它的内容如下所示。我们基本上只是删除了 `$todos` 数组和 `console.log()` 语句：

Reviewer 2: ```suggestion
4. 现在让我们删除 `App.svelte` 中硬编码的待办事项。更新它的内容如下所示。我们基本上只是删除了 `$todos` 数组和 `console.log()` 语句：
```

   ```svelte
   <script>
     import Todos from './components/Todos.svelte'
     import Alert from './components/Alert.svelte'

     import { todos } from './stores.js'
   </script>

   <Alert />
   <Todos bind:todos={$todos} />
   ```

   > **备注：** 这是我们为了使用自定义存储所必须进行的唯一更改。`App.svelte` 完全不需要知道我们使用哪种类型的 store。

5. 现在再次运行你的应用程序。创建一些待办事项，然后关闭浏览器。你甚至可以停止 Svelte 服务器并重新启动它。当你重新访问该 URL 时，你的待办事项仍将存在。
6. 你还可以在开发者工具的控制台中进行检查。在 Web 控制台中，输入命令 `localStorage.getItem('mdn-svelte-todo')`。对应用程序进行一些更改，比如按下“全部取消”按钮，然后再次检查 Web 存储的内容。你将会看到类似以下的内容：

   ![带有 Web 控制台视图的待办事项应用程序，显示当在应用程序中更改待办事项时，相应的条目也在 Web 存储中更改](03-persisting-todos-to-local-storage.png)

>    

Author: ```suggestion

```

> Svelte store 提供了一种非常简单、轻量但功能强大的方式，以响应式的方式处理来自全局数据存储的复杂应用程序状态。由于 Svelte 编译我们的代码，它可以提供 [`$store` 自动订阅语法](https://svelte.dev/docs/svelte-components#script-4-prefix-stores-with-$-to-access-their-values)，使我们可以通过与本地变量相同的方式处理存储。由于 store 具有很简单的 API，因此创建我们自己的自定义 store 来抽象 store 本身的内部工作方式非常简单。

Reviewer 2: ```suggestion
Svelte store 提供了一种非常简单、轻量但功能强大的方式，以响应式的方式处理来自全局数据存储的复杂应用程序状态。由于 Svelte 编译会我们的代码，它可以提供 [`$store` 自动订阅语法](https://svelte.dev/docs/svelte-components#script-4-prefix-stores-with-$-to-access-their-values)，使我们可以通过与本地变量相同的方式处理存储。由于 store 具有很简单的 API，因此创建我们自己的自定义 store 来抽象 store 本身的内部工作方式非常简单。
```

> Svelte store 提供了一种非常简单、轻量但功能强大的方式，以响应式的方式处理来自全局数据存储的复杂应用程序状态。由于 Svelte 编译我们的代码，它可以提供 [`$store` 自动订阅语法](https://svelte.dev/docs/svelte-components#script-4-prefix-stores-with-$-to-access-their-values)，使我们可以通过与本地变量相同的方式处理存储。由于 store 具有很简单的 API，因此创建我们自己的自定义 store 来抽象 store 本身的内部工作方式非常简单。

Author: ```suggestion
Svelte store 提供了一种非常简单、轻量但功能强大的方式，以响应式的方式处理来自全局数据存储的复杂应用程序状态。由于 Svelte 会编译我们的代码，它可以提供 [`$store` 自动订阅语法](https://svelte.dev/docs/svelte-components#script-4-prefix-stores-with-$-to-access-their-values)，使我们可以通过与本地变量相同的方式处理存储。由于 store 具有很简单的 API，因此创建我们自己的自定义 store 来抽象 store 本身的内部工作方式非常简单。
```

---
title: 使用 Svelte store
slug: Learn/Tools_and_testing/Client-side_JavaScript_frameworks/Svelte_stores
l10n:
  sourceCommit: e9be22eaa7416206e3f263a058f0d509a7f81f88
---

{{LearnSidebar}}
{{PreviousMenuNext("Learn/Tools_and_testing/Client-side_JavaScript_frameworks/Svelte_reactivity_lifecycle_accessibility","Learn/Tools_and_testing/Client-side_JavaScript_frameworks/Svelte_TypeScript", "Learn/Tools_and_testing/Client-side_JavaScript_frameworks")}}

在上一篇文章中，我们完成了应用程序的开发，将其组织为组件，并讨论了一些处理响应式、处理 DOM 节点以及暴露组件功能的高级技术。在本文中，我们将展示另一种在 Svelte 中处理状态管理的方式：[store](https://learn.svelte.dev/tutorial/writable-stores)。store 是全局数据存储库，用于保存值。组件可以订阅 store 并在其值发生变化时接收通知。

<table>
  <tbody>
    <tr>
      <th scope="row">前提：</th>
      <td>
        <p>
          至少，建议你熟悉核心的

>           <a href="/zh-CN/docs/Learn/HTML">HTML</a>、<a
>                href="/zh-CN/docs/Learn/CSS">CSS</a> 和

Reviewer 2: ```suggestion
          <a href="/zh-CN/docs/Learn/HTML">HTML</a>、<a
             href="/zh-CN/docs/Learn/CSS">CSS</a> 和
```

     <p>{ $alert }</p>
   </div>
   {/if}
   ```

这里我们首先创建了默认值为 3000（毫秒）的属性 `ms`。然后，我们创建了 `onMessageChange()` 函数，它负责控制 Alert 是否可见。通过 `$: onMessageChange($alert, ms)`，我们告诉 Svelte 在 `$alert` store 或 `ms` 属性发生变化时运行此函数。

每当 `$alert` store 发生变化时，我们将清除任何等待的超时。如果 `$alert` 为空，我们将 `visible` 设置为 `false`，之后 `Alert` 会被从 DOM 中删除。如果不为空，我们将 `visible` 设置为 `true`，并使用 `setTimeout()` 函数以在 `ms` 毫秒后清除提示。

最后，使用 `onDestroy()` 生命周期函数，确保调用 `clearTimeout()` 函数。

我们还在提示段落上方添加了 SVG 图标，使其看起来更漂亮。再次尝试一下，你应该会看到这些更改。

## 使我们的 Alert 组件具有无障碍性

我们的 `Alert` 组件工作得很好，但对辅助技术来说并不友好。问题在于页面上动态添加和删除的元素。虽然对于能看到页面的用户来说很明显，但对于使用辅助技术（如屏幕阅读器）的用户来说可能不太明显。为了处理这些情况，我们可以利用 [ARIA 实时区域](/zh-CN/docs/Web/Accessibility/ARIA/ARIA_Live_Regions)，它提供了以编程方式公开动态内容更改的方法，以便辅助技术可以检测并宣布这些更改。

> 我们可以使用 `aria-live` 属性和礼貌设置来声明包含动态内容的区域，以便辅助技术可以宣布它们。礼貌设置用于设置屏幕阅读器处理该区域更新的优先级。可能的设置有 `off`、`polite` 或 `assertive`。

Author: ```suggestion
我们可以使用 `aria-live` 属性和礼让设置来声明包含动态内容的区域，以便辅助技术可以宣布它们。礼让设置用于设置屏幕阅读器处理该区域更新的优先级。可能的设置有 `off`、`polite` 或 `assertive`。
```

> 我们可以使用 `aria-live` 属性和礼让设置来声明包含动态内容的区域，以便辅助技术可以宣布它们。礼让设置用于设置屏幕阅读器处理该区域更新的优先级。可能的设置有 `off`、`polite` 或 `assertive`。

Reviewer 2: ```suggestion
我们可以使用 `aria-live` 属性和礼让设置（politeness setting）来声明包含动态内容的区域，以便辅助技术可以宣布它们。礼让设置用于设置屏幕阅读器处理该区域更新的优先级。可能的设置有 `off`、`polite` 或 `assertive`。
```
